src/main/java/org/truffleruby/core/FinalizationService.java

```diff
@@ -124,11 +130,34 @@ private void createFinalizationThread() {
             while (true) {
                 final FinalizerReference finalizerReference = (FinalizerReference) threadManager.runUntilResult(null,
                         () -> finalizerQueue.remove());
-                finalizerReference.getFinalizerActions().forEach(action -> action.run());
+                try {
+                    finalizerReference.getFinalizerActions().forEach(action -> action.run());
+                } catch (KillException e) {
+                    throw e;
```

Comment:
`ExitException` (from `Kernel#exit!`) should not be ignored here.

Reply:
There is also `FiberShutdownException`, but that should not happen on a main thread.
Might still be safer to rethrow it as well.
I guess we should group these exceptions under a common superclass.

Reply:
Done! Especially because `FiberShutdownException` is private.